src/operator/warp.cu

```diff
@@ -0,0 +1,433 @@
+      const int gradOutputAddress = gradOutput_strideBatch * b +
+       gradOutput_strideHeight * yOut + gradOutput_strideWidth * xOut;
+
+      double topLeftDotProduct = 0;
+      double topRightDotProduct = 0;
+      double bottomLeftDotProduct = 0;
+      double bottomRightDotProduct = 0;
+
+      // we are careful with the boundaries
+      bool topLeftIsIn = between(xInTopLeft, 0, width-1) &&
+       between(yInTopLeft, 0, height-1);
+      bool topRightIsIn = between(xInTopLeft+1, 0, width-1) &&
+       between(yInTopLeft, 0, height-1);
+      bool bottomLeftIsIn = between(xInTopLeft, 0, width-1) &&
+       between(yInTopLeft+1, 0, height-1);
+      bool bottomRightIsIn = between(xInTopLeft+1, 0, width-1) &&
+       between(yInTopLeft+1, 0, height-1);
+
+      /*
+         In that loop we accumulate
+         - gradients into the gradInputImages array with atomic adds
+         - we compute the dot product that we need for the grid gradient
+      */
+
+      for (int t=threadIdx.x; t < inputImages_channels; t+= blockDim.x) {
+         double gradOutValue = gradOutput_data[gradOutputAddress + t];
+
+         // bool between(int value, int lowerBound, int upperBound)
+         if (topLeftIsIn) {
+            double inTopLeft = inputImages_data[inTopLeftAddress + t];
+            topLeftDotProduct += inTopLeft * gradOutValue;
+            if (!onlyGrid) atomicAdd(&gradInputImages_data[gradInputImagesTopLeftAddress + t],
+            xWeightTopLeft * yWeightTopLeft * gradOutValue);
+         }
+
+         if (topRightIsIn) {
+            double inTopRight = inputImages_data[inTopRightAddress + t];
+            topRightDotProduct += inTopRight * gradOutValue;
+            if (!onlyGrid) atomicAdd(&gradInputImages_data[gradInputImagesTopRightAddress + t],
+            (1 - xWeightTopLeft) * yWeightTopLeft * gradOutValue);
+         }
+
+         if (bottomLeftIsIn) {
+            double inBottomLeft = inputImages_data[inBottomLeftAddress + t];
+            bottomLeftDotProduct += inBottomLeft * gradOutValue;
+            if (!onlyGrid) atomicAdd(&gradInputImages_data[gradInputImagesBottomLeftAddress + t],
+            xWeightTopLeft * (1 - yWeightTopLeft) * gradOutValue);
+         }
+
+         if (bottomRightIsIn) {
+            double inBottomRight = inputImages_data[inBottomRightAddress + t];
+            bottomRightDotProduct += inBottomRight * gradOutValue;
+            if (!onlyGrid) atomicAdd(&gradInputImages_data[gradInputImagesBottomRightAddress + t],
+            (1 - xWeightTopLeft) * (1 - yWeightTopLeft) * gradOutValue);
+         }
+      }
+
+      /*
+         Here we reduce the dot product and compute the grid gradient before writing it.
+         could do shuffles and use no shmem at all but cuda arch is 2.0 
+      */
+
+      __shared__ volatile double __shmem[16][32];
+
+      __shmem[threadIdx.y][threadIdx.x] = topLeftDotProduct;
+      sumReduceShMem(__shmem[threadIdx.y]);

```

Comment:
May try to use the Reduce1D to replace the `sumReduceShMem` directly. 
Like to use
```c++
__syncthreads();
Reduce1D<mshadow::red::sum, warp_bits>(__shmem[threadIdx.y]);
__syncthreads();
```